docs/RFCs/0004-crashtracker-design.md

```diff
@@ -0,0 +1,146 @@
+    - Risks during a crash
+      - If the pipe/socket is not drained, the collector will stall when attempting to write.
+        This could hang the crashing process.
+        Mitigation: Minimize the amount of data sent.
+      - If the pipe/socket is not available or is closed early, data could not be sent, or the crashtracker could hang.
+        Mitigation:
+        Data is sent to the receiver in priority order.
+        If the channel is closed, the receiver will send as much data as possible, as well as a flag indicating that an error occurred.
+5.  Files collected from `/proc`
+    - Risks to normal operation
+      - N/A
+    - Risks during a crash
+      - Due to user security controls, these files may not be available to read.
+        Mitigation: The crashtracker receiver will still transmit partial crash reports.
+      - The files may be large.
+        Mitigation:
+        None currently.
+        In the future, we could add a file size limit and only send the first `n` bytes.
+      - These files could contain sensitive data
+        Mitigation: only send a select set of files which do not contain sensitive data.
+
+### Receiver
+
+A signal handler is fundamentally limited in what operations it can perform (see [signal-safety(7) \- Linux manual page](https://man7.org/linux/man-pages/man7/signal-safety.7.html)).
+This is made worse by the fact that the collector operates in the context of a crashing process whose state may be corrupted, causing even seemingly safe operations to crash/  
+The receiver interacts with the system in the following ways:
+
+1.  Forks a new process.
+    This can either occur eagerly, at process initialization or lazily, as part of crash handling.
+    Both options have issues.
+    - Risks to normal operation
+      - Eager initialization consumes a PID and a slot in the process table.
+        This can cause either initialization to fail, or block the creation of another process on the customer system.
+        Mitigation: This is unlikely to occur.
+      - Eager initialization can lead to an additional child process.
+        This was originally believed to be harmless.
+        However, it turns out that some frameworks assume that they are the only thing that can spawn workers, and do a `waitpid` on all children.
+        The existence of an additional child process hangs the framework.
+        Mitigation: Either do lazy initialization, or use a proper daemonized sidecar.
+    - Risks during a crash
+      - Lazy initialization consumes a PID and a slot in the process table.
+        This can cause either initialization to fail, or block the creation of another process on the customer system.
+        Mitigation: This is unlikely to occur.
+      - Forking inside a signal handler is unsafe (see Notes on [signal-safety(7) \- Linux manual page](https://man7.org/linux/man-pages/man7/signal-safety.7.html)).
```

Comment:
Other than at_fork handlers we need to be very careful about the code we call from within the crash handler.

We should avoid calling anything libc or any shared resource related - e.g. any internal library mutex'es must be avoided etc.

This includes most operations that allocate memory on heap- no Vec.new .etc

Anything we need to execute the handler process. must be prepared ahead of time.

Reply:
I wish you good luck with collecting stacktraces without allocations. I mean, if there's a way to do so, I'll very gladly embrace it. In fact most crashes in the crashtracker (in PHP) are just within allocations.

Reply:
> I wish you good luck with collecting stacktraces without allocations

Indeed.  Conceptually, one can imagine an unwinding strategy that uses only pre-allocated buffers.  In reality
* DWARF specifies instructions to execute in an interpreted sandbox.  It's a VM...
* Organizationally, we've gone over and over and over the idea of building an in-house remote unwinder and it keeps getting shot down.  It's a tremendous burden.  Like, pick and hand-rolled ELF parser and throw the elfutils test suite at it just to see how it crashes and burns--and that's the _easy_ part of the problem.

Reply:
This is what I'm worried about:
https://github.com/lattera/glibc/blob/master/malloc/malloc.c#L3212

There are plenty of locks throughout libc - and just like with fork(2) - accessing those from a signal handler of a crashing process is... potentially interesting.

Reply:
I think that's a strong argument in favor of collecting all the information from the receiver process, instead of in-process.

@sanchda why building our own remote unwinder instead of "just" deploying libunwind, possibly forking and customizing it for our needs?

Reply:
>I think that's a valid idea. In fact, why not just fork (in a manner that suppresses atfork handlers) and do everything in the fork (exiting the fork without calling atexit handlers), then continuing to chain in the handler after a return?

More or less 0 this is what is done in breakpad impl.

Reply:
Documented